File: src/checks/playback/tails.cpp
#include <QTest>

#include "audio/audioengine.h"
#include "checks/playback/transportfixture.h"
#include "checks/playback/tst_transport.h"
#include "core/miditimeline.h"
#include "project/voicegroupsource.h"

namespace checks {

// Stopped → Playing: the ringing tail must be cut when playback starts.
void TransportTest::playFromStoppedCutsAuditionTail()
{
    auto timeline = loadedSong(buildSilentSong(), "synthesized song built wrong");
    QVERIFY2(timeline, "synthesized song built wrong");
    engine().loadSong(timeline, borrowVoicegroupLease(&m_bank->vg), SongSettings{});

    if (!ringingTail(engine(), 0))
        return;
    engine().play();
    QVERIFY2(QTest::qWaitFor([&] { return engine().activePcmChannels() == 0; }, 2000),
             "audition tail persisted after playback started from stop");
}

// Playing → Paused: pausing silences like Stop — a preview sounding when
// pause hits must not ring through it.
void TransportTest::pauseSilencesPlayingPreview()
{
    auto timeline = loadedSong(buildSilentSong(), "synthesized song built wrong");
    QVERIFY2(timeline, "synthesized song built wrong");
    engine().loadSong(timeline, borrowVoicegroupLease(&m_bank->vg), SongSettings{});
    engine().play();
    QVERIFY2(QTest::qWaitFor([&] { return engine().playheadSamples() > 0; }, 2000),
             "playback did not start for the pause check");

    engine().previewNoteTimed(0, 60, 127, uint32_t(60.0 * engine().sampleRate()));
    QVERIFY2(QTest::qWaitFor([&] { return engine().activePcmChannels() >= 1; }, 2000),
             "timed preview during playback never sounded");
    engine().pause();
    QVERIFY2(QTest::qWaitFor([&] { return engine().activePcmChannels() == 0; }, 2000),
             "pause left the preview ringing");
}

// Paused → Playing, the Space path (pause, audition, seek + play): Space
// toggles pause and restarts from the edit cursor, so this is how playback
// usually starts — the tail must be cut here too.
void TransportTest::spacePathSeekAndPlayCutsTail()
{
    auto timeline = loadedSong(buildSilentSong(), "synthesized song built wrong");
    QVERIFY2(timeline, "synthesized song built wrong");
    engine().loadSong(timeline, borrowVoicegroupLease(&m_bank->vg), SongSettings{});
    engine().play();
    QVERIFY2(QTest::qWaitFor([&] { return engine().playheadSamples() > 0; }, 2000),
             "playback did not start for the Space-path check");
    engine().pause();
    QTest::qWait(300); // let the pause transition settle before auditioning

    if (!ringingTail(engine(), 0))
        return;
    engine().seek(0);
    engine().play();
    QVERIFY2(QTest::qWaitFor([&] { return engine().activePcmChannels() == 0; }, 2000),
             "audition tail persisted after Space-style seek + play from pause");
}

// Paused → Playing with the preview still counting down: resuming must cut
// it rather than let it sound over the song for the full duration.
void TransportTest::resumeCutsCountingDownPreview()
{
    auto timeline = loadedSong(buildSilentSong(), "synthesized song built wrong");
    QVERIFY2(timeline, "synthesized song built wrong");
    engine().loadSong(timeline, borrowVoicegroupLease(&m_bank->vg), SongSettings{});
    engine().play();
    QVERIFY2(QTest::qWaitFor([&] { return engine().playheadSamples() > 0; }, 2000),
             "playback did not start for the resume check");
    engine().pause();
    QTest::qWait(300); // let the pause transition settle before auditioning

    engine().previewNoteTimed(1, 64, 127, uint32_t(60.0 * engine().sampleRate()));
    QVERIFY2(QTest::qWaitFor([&] { return engine().activePcmChannels() >= 1; }, 2000),
             "timed preview during pause never sounded");
    engine().play();
    QVERIFY2(QTest::qWaitFor([&] { return engine().activePcmChannels() == 0; }, 2000),
             "counting-down preview persisted after resuming playback");
}

// Unload with the song still playing — the song-switch path. unloadSong
// assigns both transport fields itself, so the callback never sees a
// Playing→Stopped transition and no transport cut-fade ever runs there; the
// cut must happen inside unloadSong. MainWindow::loadSong frees the
// outgoing voicegroup right after unloadSong returns, so freeAll + the wait
// below give ASAN a window to catch any channel still rendering it.
void TransportTest::unloadWhilePlayingCutsSongVoices()
{
    HeapVoicegroup hvg;
    auto timeline = loadedSong(buildNoteSong(), "note song built wrong");
    QVERIFY2(timeline, "note song built wrong");
    QVERIFY2(timeline->usedTrackCount == 1, "note song built wrong");
    engine().loadSong(timeline, borrowVoicegroupLease(hvg.vg), SongSettings{});
    engine().play();
    if (!QTest::qWaitFor([&] { return engine().activePcmChannels() >= 1; }, 2000)) {
        engine().unloadSong();
        QTest::qFail("song note never sounded before unload", __FILE__, __LINE__);
        return;
    }
    engine().unloadSong();
    QVERIFY2(QTest::qWaitFor([&] { return engine().activePcmChannels() == 0; }, 2000),
             "unloadSong left song channels sounding — they outlive the voicegroup "
             "the switch path frees");

    hvg.freeAll();
    // The device keeps calling back; a channel that survived the unload now
    // reads the freed WaveData (ASAN reports it even when the count above
    // somehow reached zero).
    QTest::qWait(300);
}

} // namespace checks

File: src/checks/playback/tails.swift
import Foundation
@testable import PorydawApp
import PorydawCore
import PorydawPlayback
import PorydawPlaybackNative

func checkAuditionTailTransitions(_ report: CheckReport) throws {
    for scenario in 0..<4 {
        let names = ["playFromStoppedCutsAuditionTail", "pauseSilencesPlayingPreview",
                     "spacePathSeekAndPlayCutsTail", "resumeCutsCountingDownPreview"]
        let id = "transportcheck/TransportTest::\(names[scenario])"
        let rig = try AudioControllerCheckFixture(square: true, silent: true)
        for program in 0..<128 { rig.voices[program].release = 254 }
        let audio = rig.renderer
        let timeline = playbackCheckSilentTimeline()
        audio.bind(timeline: timeline, voicegroup: rig.voices, settings: AudioSettings())
        report.expectEqual(2, timeline.usedTrackCount, cppID: id, what: "both audition tracks initialized")
        report.expect(audio.songLoaded, cppID: id, message: "silent song loaded")
        if scenario != 0 {
            audio.play()
            _ = rig.render(rig.rate / 4)
            report.expect(audio.playheadSamples > 0, cppID: id, message: "playback started")
            if scenario != 1 {
                audio.pause()
                _ = rig.render(rig.rate * 3 / 10)
            }
        }
        let tail = scenario == 0 || scenario == 2
        audio.audition.previewNoteTimed(track: scenario == 3 ? 1 : 0,
            key: scenario == 3 ? 64 : 60, velocity: 127,
            durationSamples: UInt32(tail ? rig.rate * 15 / 100 : rig.rate * 60))
        _ = rig.render(512)
        report.expect(audio.activePcmChannels >= 1, cppID: id,
                      message: "timed preview sounds before transition")
        if tail {
            _ = rig.render(rig.rate * 4 / 10)
            report.expect(audio.polySnapshot().pcm.contains { $0.on && $0.releasing },
                cppID: id, message: "expired preview has a live release tail")
        }
        if scenario == 1 {
            audio.pause()
        } else {
            if scenario == 2 { audio.seek(0) }
            audio.play()
        }
        _ = rig.render(rig.rate * 2)
        report.expectEqual(0, audio.activePcmChannels, cppID: id,
                          what: "transition cuts preview and release tail within two seconds")
    }
    try checkUnloadedVoicegroupLifetime(report)
}

func checkUnloadPlayingSong(_ rig: AudioControllerCheckFixture, _ report: CheckReport) {
    let audio = rig.renderer
    report.expect(audio.activePcmChannels >= 1,
        cppID: "transportcheck/TransportTest::unloadWhilePlayingCutsSongVoices",
        message: "song voice sounds before unload")
    audio.unload()
    _ = rig.render(8192)
    report.expect(audio.activePcmChannels == 0 && audio.activeCgbChannels == 0 && !audio.songLoaded,
        cppID: "transportcheck/TransportTest::unloadWhilePlayingCutsSongVoices", message: "unload retires voices before bank release")
}

private func checkUnloadedVoicegroupLifetime(_ report: CheckReport) throws {
    let id = "transportcheck/TransportTest::unloadWhilePlayingCutsSongVoices"
    var bank: AudioControllerCheckFixture? = try AudioControllerCheckFixture(square: true)
    let audio = bank!.renderer
    report.expect(audio.songLoaded, cppID: id, message: "heap-bank note song loaded")
    report.expectEqual(1, audio.timeline!.usedTrackCount, cppID: id, what: "one song track")
    audio.play()
    _ = bank!.render(12_000)
    report.expect(audio.activePcmChannels >= 1, cppID: id, message: "heap-bank note sounds")
    audio.unload()
    report.expectEqual(0, audio.activePcmChannels, cppID: id, what: "unload immediately retires song voices")
    bank = nil
    var output = [Float](repeating: 0, count: 1024)
    for _ in 0..<29 {
        output.withUnsafeMutableBufferPointer { audio.render($0.baseAddress!, frames: 512) }
    }
    report.expectEqual(0, audio.activePcmChannels, cppID: id,
                      what: "callbacks remain voiceless after heap-bank destruction")
}
